Add swap_tensors path to nn.Module._apply

mikaylagawarecki opened the pull request:
Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run*** if the autograd graph is still alive. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap the refs that the `AccumulateGrad` nodes hold if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad` OR the autograd graph is no longer alive because the output has been garbage collected.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118023
* __->__ #117167
* #118028

mikaylagawarecki committed:
Add swap_tensors path to nn.Module._apply

[ghstack-poisoned]

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -823,26 +831,48 @@ def compute_should_use_set_data(tensor, tensor_applied):
-                out_param = param
+            p_should_use_set_data = compute_should_use_set_data(param, param_applied)
+            p_should_use_swap_tensors = compute_should_use_swap_tensors(param)
+            # FIXME: This is wrong, param_grad bumps the use_count so we will never swap_tensors for grad
+            # nevertheless, passing param.grad to any function (even swap_tensors) bumps its use_count,
+            # so it seems like we are maybe doomed?
```
@albanD 

I am running into an issue with the `use_count` check we added, this happens specifically for `.grad`.

Consider the following

```python
>>> import torch
>>> t = torch.randn(2, 3)
>>> p = torch.nn.Parameter(t)
>>> p.grad = torch.randn(2, 3)
>>> torch._C._tensor_use_count(p.grad._cdata)
1
>>> def foo(g):
>>>     print(torch._C._tensor_use_count(g._cdata)
>>> foo(p.grad) # 2
```

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792




[ghstack-poisoned]

albanD reviewed (commented):
I don't think that falling back to .data= is really an option here. They are doing semantically different things and so we should only switch between the two when the user explicitly asks for it.

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify `nn.Module._apply. compute_should_use_swap_tensors` to override this if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). Future work might be to swap `AccumulateGrad` nodes if it is necessary.

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

### Question
If the future is set to allow swapping, there are still cases where swapping might not occur (`use_count > 1`, `weak_use_count > 1` (`weak_use_count` is `(use_count >= 1) + num_weak_refs`).

I am wondering what we should do in such cases:
1) error loudly: 
    - Pro: for use cases where `swap_tensors` is necessary for correctness (`XLATensor`, `DTensor`), this will make it very apparent that things have gone wrong
    - Con: For other use cases where `.data` setting is not semantically correct perhaps this might not preserve BC, especially if we flip the default (before weakrefs to parameters were ok, now not anymore)
2) Warn and fall back to `.data` setting: I was thinking to warn with a list of param_names that were not swapped (since only some might have weakrefs, use_count > 1 etc.)
     - Pro: Doesn't break BC for the common case, provides signal for how to fix for the case where `._apply` is currently broken
     - Con: warning might be spammy, especially if there are a lot of parameters
3) silently fall back to `.data` setting 
     - Pro: Not spammy
     - Con: V hard to debug correctness

Right now I haven't chosen which to implement yet, so the fallback is just silent.I am wondering which of (1) or (2) (or something else) might be the better solution?





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify `nn.Module._apply. compute_should_use_swap_tensors` to override this if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). Future work might be to swap `AccumulateGrad` nodes if it is necessary.

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

### Question
If the future is set to allow swapping, there are still cases where swapping might not occur (`use_count > 1`, `weak_use_count > 1` (`weak_use_count` is `(use_count >= 1) + num_weak_refs`).

I am wondering what we should do in such cases:
1) error loudly: 
    - Pro: for use cases where `swap_tensors` is necessary for correctness (`XLATensor`, `DTensor`), this will make it very apparent that things have gone wrong
    - Con: For other use cases where `.data` setting is not semantically correct perhaps this might not preserve BC, especially if we flip the default (before weakrefs to parameters were ok, now not anymore)
2) Warn and fall back to `.data` setting: I was thinking to warn with a list of param_names that were not swapped (since only some might have weakrefs, use_count > 1 etc.)
     - Pro: Doesn't break BC for the common case, provides signal for how to fix for the case where `._apply` is currently broken
     - Con: warning might be spammy, especially if there are a lot of parameters
3) silently fall back to `.data` setting 
     - Pro: Not spammy
     - Con: V hard to debug correctness

Right now I haven't chosen which to implement yet, so the fallback is just silent.I am wondering which of (1) or (2) (or something else) might be the better solution?





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap `AccumulateGrad` nodes if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad`.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

ysiraichi commented:
@mikaylagawarecki Hi, there. Just wanted to give a heads-up that I opened #118783, a complementary PR which also fixes #115792. While it fixes the same issue, it does so because of another latent problem with `FunctionalTensorWrapper`. Check [this comment](https://github.com/pytorch/pytorch/issues/115792#issuecomment-1913311298) to read more about it. That said, I believe this PR is still relevant, since it also addresses other issues.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -823,26 +825,47 @@ def compute_should_use_set_data(tensor, tensor_applied):
+                    if param_grad is not None:
+                        # Accessing param.grad makes its use_count 2, which will prevent swapping.
+                        # Decrement use count of the gradient by setting to None
+                        param.grad = None
```
I would move this into the try and add the setting it back in a finally: block

albanD commented on torch/nn/modules/module.py:
```diff
@@ -823,26 +825,47 @@ def compute_should_use_set_data(tensor, tensor_applied):
+                    try:
+                        torch.utils.swap_tensors(param, param_applied)
+                    except Exception as e:
+                        raise RuntimeError(f"Couldn't swap {key}") from e
```
nit: should this be more precise about while moving module parameter or something? The key here will only contain the key within this module, not the fqn!

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -823,26 +825,47 @@ def compute_should_use_set_data(tensor, tensor_applied):
                 else:
-                    assert param.grad.is_leaf
-                    out_param.grad = grad_applied.requires_grad_(param.grad.requires_grad)
+                    assert param_grad.is_leaf
```
This is copied from existing code, I wasn't sure on the rationale for this either 🤔

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.nn.utils.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is on XLA or is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run***. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap the refs that the `AccumulateGrad` nodes hold if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad` OR the autograd graph is no longer alive because the output has been garbage collected.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki commented on test/test_modules.py:
```diff
@@ -853,6 +853,79 @@ def test_errors(self, device, dtype, module_info, training):
+
+                m = module_cls(*c_args, **c_kwargs)
+
+                def _to(m, set_grad=False):
```
Hm this is not really "old-style" the tests in `test_modules` always take in a device and dtype, the first step is usually to move the constructed module to device/dtype using `module.to()`. I am trying to avoid using `module.to()` (whether swapping is enabled or disabled)  since that is the method that this test is intended to test, added a comment

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run*** if the autograd graph is still alive. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap the refs that the `AccumulateGrad` nodes hold if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad` OR the autograd graph is no longer alive because the output has been garbage collected.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki commented on torch/__future__.py:
```diff
@@ -1,21 +1,60 @@
+    #. ``module.{device}()`` (e.g. ``module.cuda()``) for moving a module between devices
+    #. ``module.{dtype}()`` (e.g. ``module.float()``) for converting a module to a different dtype
+       (for converting a module to a different dtype)
+    #. ``module.to()``
```
should we mention `_apply()`? I omitted it since it is a private method

albanD replied:
No we shouldn't talk about it.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -823,26 +825,47 @@ def compute_should_use_set_data(tensor, tensor_applied):
+                            raise RuntimeError(f"Couldn't swap {key}.grad") from e
+                        out_param.grad = param_grad
+                    else:
+                        assert out_param.grad is not None
```
Ok ok!

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run*** if the autograd graph is still alive. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap the refs that the `AccumulateGrad` nodes hold if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad` OR the autograd graph is no longer alive because the output has been garbage collected.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add swap_tensors path to nn.Module._apply"

Towards fixing https://github.com/pytorch/pytorch/issues/115792

Added `torch.__future__.{get/set}_swap_module_params_on_conversion` that defaults to `False` for now, but we probably want to modify  to override this and default to `True` in `nn.Module._apply` if input is a tensor subclass.

From offline discussion, for now we are **not** allowing `swap_tensor` after the first module forward has been run*** if the autograd graph is still alive. The reason being that `torch.utils.swap_tensors(t1, t2)` requires the `use_count` of both `TensorImpl`s associated with `t1` and `t2` to be 1.  The first forward pass will install `AccumulateGrad` nodes on each param, which [bump the refcount of the associated TensorImpl](https://github.com/pytorch/pytorch/blob/6cf1fc66e340132d7e2ed9d42efea42fa7ea0183/torch/csrc/autograd/variable.cpp?fbclid=IwAR2dWDVPoXfWF0QDXhhwJ3U7CIAUcNBCAxptlTX9yDI-0pi_h0FBNsw0ig0#L307). **Future work might be to swap the refs that the `AccumulateGrad` nodes hold if it is necessary.**

***From this, it might seem like we don't need to handle gradients. However, I still handle the grads for the edge case that the grads are set via `p.grad = grad` OR the autograd graph is no longer alive because the output has been garbage collected.

If any `swap_tensors` fails on any of the parameters in the `nn.Module` we raise an error.


**`RNNBase` overrides `nn.Module._apply()` and installs weakrefs on some parameters. As a result, all modules that inherit from `RNNBase` (`RNN`, `GRU` and `LSTM`) cannot use the`swap_tensors` path as of now**





[ghstack-poisoned]

albanD reviewed (approved):
Thanks!
